Refactor the layout conceptual section (#76)
pavpanchekha/emberfox: book/layout.md

we'll add support for backgrounds to make our web pages more colorful.

The Layout Tree
===============

The way our browser works now, laying out an element involves separate
modifications to global state at the open tag, for each child, and
then once more for the close tag. While this does put things in the
right place, information about the element as a whole, like its width
and height, isn't computed. That makes it pretty hard to draw a
background, without knowing how wide and tall to draw it.

So web browsers structure layout differently. In a browser, layout
produces a *layout tree* of layout objects associated with the HTML
elements.[^no-box] Each of those layout objects has a size and a
position, and the layout process is thought of as generating the
layout tree and then computing those sizes and positions.

[^no-box]: Some elements like `<script>` don't generate layout
    objects, and some elements like `<li>` generate multiple (one for
    the bullet point!), but for most elements it's one element one
    layout object.

For example, consider a web page with a body that contains a heading
and three paragraphs:

>     <body>

Reviewer: I think it would be great, and instructive, to make this a real example page, perhaps in an iframe so that can also be loaded on its own. I will try this out in a PR once you've landed this one.

Author: This is a great idea!

one layout object inside that for each heading and paragraph. Each
layout object will store its size and position, so for example if the
body element has a background the body element's size and position can
be used to draw that background.

The layout tree drives the rest of the layout process. The browser
first traverses the HTML tree to create the layout tree. Then it
computes the size and position for each layout object. Finally, each
layout object generates drawing commands which the browser executes
to render the page.

Layout modes
============

Different layout object lay out their children differently. Some
elements contain text, like the paragraph above, and lay that text out
horizontally in lines. But other elements contain those containers
themselves. In the example bove, the body element contains a heading
and some paragraphs, which are themselves containers for text. Those
containers are stacked vertically, from top to bottom. Abstracting a
bit, there are two *layout modes* here, two ways an element can be
laid out relative to its children.[^or-equivalently] They're called
block layout, for laying out containers, and inline layout, for laying

> out text, so we say the body element is in block layout mode (because

Reviewer: ```suggestion
out text.

 In the example, the body element is in block layout mode (because
```

its children are containers) while the paragraphs and heading are in
inline layout mode (because their children are text). In real browsers
there are a lot of other layout modes too, but in this chapter we'll
keep it simple.

[^or-equivalently]: Or, equivalently, how an element is laid out
relative to its siblings and parent. CSS tends to this equivalent
perspective, though in some cases (like the `inline-block` display
mode) it also defines the relationship to children.

> Let's have one class for each layout mode. Inline layout should be

Reviewer: ```suggestion
Let's have one class for each layout mode. InlineLayout should be
```

familiar: it's what we've been implementing up until this point and
corresponds to how text inside a paragraph are laid out in lines, left

> to right.[^in-english] Block layout is new to this chapter: it

Reviewer: ```suggestion
to right.[^in-english] BlockLayout is new to this chapter: it
```

    self.y = self.parent.y
```

These three computations have to go before the loop that calls
`layout` on each child. After all, a block layout object's width

> depends on the parent's width; so by symmetry the width must be

Reviewer: ```suggestion
depends on the parent's width; so the width must be
```
